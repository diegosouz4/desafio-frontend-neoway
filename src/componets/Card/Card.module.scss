.card {
  padding: 0.625rem;
  flex: 0 0 300px;

  opacity: 0;
  transform: translate3d(0, 50px, 0);
  position: relative;
  animation: 200ms fadeInUp 1s forwards ease;

  .wrapper {
    background-color: var(--color-white-100);
    border-radius: 0.25rem;
    box-shadow: 0 0.5rem 1rem -0.25rem rgba(0, 0, 0, 0.2);
    height: 100%;
    overflow: hidden;

    position: relative;
    transition: 450ms var(--pop);

    &:not(:hover, :focus) {
      opacity: 0.85;
    }

    &:hover,
    &:focus {
      opacity: 1;
      transform: scale(1.02);
    }

    display: grid;
    grid-template-rows: auto 120px;
    gap: 1rem;

    .likes {
      background: none;
      background-color: rgba(255, 255, 255, 0.85);
      border: 1px solid rgba(255, 255, 255, 0.25);
      border-radius: 50%;
      box-shadow: 0 0.5rem 1rem -0.25rem rgba(0, 0, 0, 0.35);
      width: 2rem;
      height: 2rem;

      display: flex;
      align-items: center;
      justify-content: center;

      position: absolute;
      top: 0.625rem;
      right: 0.625rem;
      z-index: 1;

      &:focus {
        .icon {
          animation: puse 800ms forwards ease;
        }
      }

      .icon {
        font-size: 1rem;
        color: var(--color-main-200);

        position: relative;
        transform: scale3d(1, 1, 1);

        &.fill {
          color: var(--color-main);
        }
      }
    }

    .thumb {
      display: block;
      width: 100%;
      height: 150px;
      background: var(--card-thumb) url() no-repeat center center;
      background-size: cover;
      position: relative;

      &::before {
        content: "";
        display: block;
        width: 100%;
        height: 100%;
        background: var(--card-overlay);
        mix-blend-mode: screen;

        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
      }
    }

    .content {
      color: var(--color-text);
      padding: 0 0.75rem 0.625rem;
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      .title {
        font-size: 1.125rem;
        font-weight: 600;
        line-height: 1.25;
      }

      .info {
        border-top: 1.5px solid #c5c5c5;
        padding-top: 0.625rem;

        display: flex;
        align-items: center;
        justify-content: space-between;
        span {
          color: var(--card-icons);
          display: block;
          font-size: 0.875rem;
          font-weight: 400;
        }
      }
    }
  }
}

@keyframes fadeInUp {
  from {
    opacity: 0;
    transform: translate3d(0, 50px, 0);
  }
  to {
    opacity: 1;
    transform: translate3d(0, 0, 0);
  }
}

@keyframes puse {
  0% {
    color: currentColor;
    transform: scale3d(0.5, 0.5, 0.5);
  }
  50% {
    color: var(--color-heart);
    transform: scale3d(1.4, 1.4, 1.4);
  }
  75% {
    color: currentColor;
    transform: scale3d(1, 1, 1);
  }
}
