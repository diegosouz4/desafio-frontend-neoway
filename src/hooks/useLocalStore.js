export function SetLocalStore(key, state) {
  if (key !== undefined) {
    window.localStorage.setItem(key, JSON.stringify(state));
    return state;
  }

  return null;
}

export function GetLocalStore(key) {
  const valueLocalStore = window.localStorage.getItem(key);
  if (!key || !valueLocalStore) return null;

  return JSON.parse(valueLocalStore);
}
