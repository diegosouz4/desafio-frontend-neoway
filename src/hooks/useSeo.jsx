export default function useSeo(title, description) {
  if (title !== undefined && title !== null) {
    document.title = title;
  }

  if (description !== undefined && description !== null) {
    document.querySelector("meta[name='description']").content = description;
  }

  return [title, description];
}
