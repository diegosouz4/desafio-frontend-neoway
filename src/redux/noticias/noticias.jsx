import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  news: {
    loading: false,
    data: null,
    error: null,
  },
};

const slice = createSlice({
  name: "noticias",
  initialState: initialState,
  reducers: {
    fetchNewsStart(state) {
      (state.news.loading = true),
        (state.news.data = null),
        (state.news.error = null);
    },
    fetchNewsSuccess(state, action) {
      (state.news.loading = false),
        (state.news.data = action.payload),
        (state.news.error = null);
    },
    fetchNewsError(state, action) {
      (state.news.loading = false),
        (state.news.data = null),
        (state.news.error = action.payload);
    },
  },
});

const { fetchNewsStart, fetchNewsSuccess, fetchNewsError } = slice.actions;

export const fetchNoticias = (query) => async (dispatch) => {
  const { q, language, sortBy } = query;
  const url = `https://newsapi.org/v2/everything?q=${q}&language=${language}&sortBy=${sortBy}&from=2023-09-24&apiKey=${
    import.meta.env.VITE_API_TOKEN_TESTE
  }&pageSize=100`;

  try {
    dispatch(fetchNewsStart());
    const response = await fetch(url);
    const data = await response.json();
    //Removendo as notícias
    data.articles = await data.articles.filter(
      (item) => item.title !== "[Removed]" && item.title !== null
    );
    //Alterando os id's null pelo index da notícia.
    data.articles = await data.articles.map((item, index) =>({...item, source: {...item.source, id: index}}));
    return dispatch(fetchNewsSuccess(data));
  } catch (error) {
    dispatch(fetchNewsError(error.message));
  }
};

export default slice.reducer;
