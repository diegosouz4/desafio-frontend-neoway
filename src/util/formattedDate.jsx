export default function formattedDate(date) {
  if (!date) return;

  const newDate = date.split("T")[0].replaceAll("-", "/");
  return newDate;
}
