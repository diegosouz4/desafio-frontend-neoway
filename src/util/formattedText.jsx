export default function formattedText(text, limit = 50) {
  if (!text || text.length < limit) return text;
  let formattedText = text.substring(0, limit);
  formattedText = `${formattedText}...`;
  return formattedText;
}
