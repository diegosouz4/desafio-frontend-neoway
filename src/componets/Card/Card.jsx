import formattedText from "../../util/formattedText";
import style from "./Card.module.scss";
import PropTypes from "prop-types";
import PostInfo from "../PostInfo/PostInfo";
import { Link } from "react-router-dom";
import { SetLocalStore } from "../../hooks/useLocalStore";
import { FaHeart, FaRegHeart } from "react-icons/fa6";
import { useDispatch } from "react-redux";
import { addLikesItem } from "../../redux/noticias/noticias";

export default function Card({ noticia }) {
  const title = formattedText(noticia?.title);
  const dispatch = useDispatch();

  const handleClick = () => {
    SetLocalStore("post", noticia);
  };

  const handleLikesClick = (e) => {
    e.preventDefault();
    dispatch(addLikesItem(noticia));
  };

  return (
    <li className={style.card}>
      <Link
        className={style.wrapper}
        to={`/noticia/${noticia.title}`}
        onClick={handleClick}
      >
        <div
          className={style.thumb}
          style={{ backgroundImage: `url(${noticia?.urlToImage})` }}
        ></div>
        <div className={style.content}>
          <h2 className={style.title}>{title}</h2>

          <PostInfo
            publishedAt={noticia?.publishedAt}
            author={noticia?.author}
            className={style.info}
          />
        </div>

        <button className={style.likes} onClick={handleLikesClick}>
          {noticia.likes ? (
            <>
              {" "}
              <span className="sronly"> remover dos favorito</span>
              <FaHeart className={style.icon} />
            </>
          ) : (
            <>
              {" "}
              <span className="sronly">salvar nos favorito</span>
              <FaRegHeart className={style.icon} />
            </>
          )}
        </button>
      </Link>
    </li>
  );
}

Card.propTypes = {
  noticia: PropTypes.object.isRequired,
};
