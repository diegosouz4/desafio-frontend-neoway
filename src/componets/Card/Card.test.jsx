import "@testing-library/jest-dom";
import { render, fireEvent, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import Card from "./Card";
import formattedText from "../../util/formattedText";
import formattedDate from "../../util/formattedDate";
import { addLikesItem } from "../../redux/noticias/noticias";
import { SetLocalStore } from "../../hooks/useLocalStore";

const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

jest.mock("../../redux/noticias/noticias", () => {
  return { addLikesItem: jest.fn() };
});

jest.mock("../../hooks/useLocalStore", () => {
  return { SetLocalStore: jest.fn() };
});

const mockStore = configureStore([]);

describe("<Card />", () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      noticias: {},
    });
  });

  it("should be render with the props passed", () => {
    const mockNoticia = {
      title: "Sample Title",
      urlToImage: "sample-image.jpg",
      publishedAt: "2023-10-05",
      author: "Jest",
      likes: false,
    };

    mockNoticia.title = formattedText(mockNoticia.title);
    mockNoticia.publishedAt = formattedDate(mockNoticia.publishedAt);

    const { asFragment, getByText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <Card noticia={mockNoticia} />
        </Provider>
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();

    expect(getByText("Sample Title")).toBeInTheDocument();
    expect(getByText("2023/10/05")).toBeInTheDocument();
    expect(getByText("Jest")).toBeInTheDocument();
  });

  it("should the dispath of the like button is passing the news ", async () => {
    const mockNoticia = {
      title: "Sample Title",
      likes: false,
    };

    const { getByText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <Card noticia={mockNoticia} />
        </Provider>
      </BrowserRouter>
    );

    const likesBtn = getByText("salvar nos favorito");
    expect(likesBtn).toBeInTheDocument();

    await fireEvent.click(likesBtn);

    expect(addLikesItem).toHaveBeenCalledTimes(1);
    expect(addLikesItem).toHaveBeenCalledWith(mockNoticia);
  });

  it("shoul view the news when clicked", () => {
    const mockNoticia = {
      title: "News title",
    };

    render(
      <BrowserRouter>
        <Provider store={store}>
          <Card noticia={mockNoticia} />
        </Provider>
      </BrowserRouter>
    );

    const link = screen.getByRole("link");

    expect(link).toBeInTheDocument();
    expect(link.getAttribute('href')).toBe("/noticia/News title");

    fireEvent.click(link);

    expect(SetLocalStore).toHaveBeenCalledWith('post', mockNoticia);
  });
});
