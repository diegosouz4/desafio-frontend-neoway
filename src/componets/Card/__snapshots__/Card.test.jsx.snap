// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should be render with the props passed 1`] = `
<DocumentFragment>
  <li>
    <a
      href="/noticia/Sample Title"
    >
      <div
        style="background-image: url(sample-image.jpg);"
      />
      <div>
        <h2>
          Sample Title
        </h2>
        <div>
          <span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
              />
            </svg>
             2023/10/05
          </span>
          <span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
              />
            </svg>
             Jest
          </span>
        </div>
      </div>
      <button>
         
        <span
          class="sronly"
        >
          salvar nos favorito
        </span>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M225.8 468.2l-2.5-2.3L48.1 303.2C17.4 274.7 0 234.7 0 192.8v-3.3c0-70.4 50-130.8 119.2-144C158.6 37.9 198.9 47 231 69.6c9 6.4 17.4 13.8 25 22.3c4.2-4.8 8.7-9.2 13.5-13.3c3.7-3.2 7.5-6.2 11.5-9c0 0 0 0 0 0C313.1 47 353.4 37.9 392.8 45.4C462 58.6 512 119.1 512 189.5v3.3c0 41.9-17.4 81.9-48.1 110.4L288.7 465.9l-2.5 2.3c-8.2 7.6-19 11.9-30.2 11.9s-22-4.2-30.2-11.9zM239.1 145c-.4-.3-.7-.7-1-1.1l-17.8-20c0 0-.1-.1-.1-.1c0 0 0 0 0 0c-23.1-25.9-58-37.7-92-31.2C81.6 101.5 48 142.1 48 189.5v3.3c0 28.5 11.9 55.8 32.8 75.2L256 430.7 431.2 268c20.9-19.4 32.8-46.7 32.8-75.2v-3.3c0-47.3-33.6-88-80.1-96.9c-34-6.5-69 5.4-92 31.2c0 0 0 0-.1 .1s0 0-.1 .1l-17.8 20c-.3 .4-.7 .7-1 1.1c-4.5 4.5-10.6 7-16.9 7s-12.4-2.5-16.9-7z"
          />
        </svg>
      </button>
    </a>
  </li>
</DocumentFragment>
`;
